File: src/screens/instructions.c
#include "raylib.h"
#include "screens.h"
#include <stdio.h>

static GameScreen finishScreen = -1;
static Font font = {0};
static Font instructionsFont = {0};
static Vector2 mousePosition = {0.0f, 0.0f};
static const Color backgroundColor = (Color){47, 79, 146, 255};

static Vector2 textPosition = {0};
static const char *text = "Instructions";
static const int titleFontSize = 72;

static char *playText = "Play";
static Vector2 playTextPosition = {0};
static Vector2 playTextSize = {0};
static const int supplementalFontSize = 36;

// Instructions Screen Initialization logic
void InitInstructionsScreen(void) {
    SetMouseCursor(MOUSE_CURSOR_DEFAULT);

    font = LoadFontEx("assets/fonts/hello-santa.ttf", 96, 0, 0);
    instructionsFont = LoadFontEx("assets/fonts/coolvetica.ttf", 96, 0, 0);

    GenTextureMipmaps(&font.texture);
    SetTextureFilter(font.texture, TEXTURE_FILTER_POINT);
    const Vector2 textSize = MeasureTextEx(font, text, titleFontSize, 1);
    textPosition = (Vector2){(GetScreenWidth() - textSize.x) / 2.0f, 10.0f};

    playTextSize =
        MeasureTextEx(GetFontDefault(), playText, supplementalFontSize, 1);
    playTextPosition =
        (Vector2){(GetScreenWidth() - playTextSize.x) / 2.0f,
                  (GetScreenHeight() - playTextSize.y) / 2.0f + 150};
    finishScreen = -1;
}

// Instructions Screen Update logic
void UpdateInstructionsScreen(void) {
    SetMouseCursor(MOUSE_CURSOR_DEFAULT);
    mousePosition = GetMousePosition();

    if (CheckCollisionPointRec(
            mousePosition, (Rectangle){playTextPosition.x, playTextPosition.y,
                                       playTextSize.x, playTextSize.y})) {
        SetMouseCursor(MOUSE_CURSOR_POINTING_HAND);
        if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON)) {
            finishScreen = GAMEPLAY;
        }
    }
}

// Instructions Screen Draw logic
void DrawInstructionsScreen(void) {
    DrawRectangle(0, 0, GetScreenWidth(), GetScreenHeight(), backgroundColor);
    DrawTextEx(font, text, textPosition, titleFontSize, 1,
               (Color){255, 255, 255, 255});
    const int fontSize = 24;
    const int startPositionY = 200;
    // Draw actual instructions
    DrawTextEx(instructionsFont,
               "Hold down H or O to make Santa run. HO, HO, HO!",
               (Vector2){10, startPositionY}, fontSize, 1, WHITE);
    DrawTextEx(instructionsFont, "Release the key to stop him.",
               (Vector2){10, startPositionY + 40}, fontSize, 1, WHITE);
    DrawTextEx(
        instructionsFont,
        "Stop as close as possible to the candy cane, but don't touch it "
        "or you know what happens!",
        (Vector2){10, startPositionY + 80}, fontSize, 1, WHITE);
    DrawTextEx(instructionsFont,
               "The closer you get to the candy cane, the better, so try for "
               "your lowest score!",
               (Vector2){10, startPositionY + 120}, fontSize, 1, WHITE);
    DrawTextEx(instructionsFont, playText, playTextPosition,
               supplementalFontSize, 1, WHITE);
}

// Instructions Screen Unload logic
void UnloadInstructionsScreen(void) {
    UnloadFont(font);
    UnloadFont(instructionsFont);
}

// Instructions Screen should finish?
GameScreen FinishInstructionsScreen(void) { return finishScreen; }

File: love2d/states/instructions.lua
local GameStateManager = require("lib/GameStateManager")
local Assets = require("assets")

local Instructions = {}

function Instructions:new()
  local obj = {
    title = "Instructions",
    play = "Play",
    instructions_start_y = 250,
    hand_cursor = love.mouse.getSystemCursor("hand"),
  }
  setmetatable(obj, self)
  self.__index = self
  return obj
end

function Instructions:enter()
  love.mouse.setCursor()

  self.font = Assets.fonts.title
  self.menu_font = Assets.fonts.menu
  self.menu_font:setFilter("nearest", "nearest")
  self.santa_walk = Assets.sprites.santa_walk
  self.background_color = Assets.colors.background
  self.background = Assets.sprites.background
  self.instructions_font = love.graphics.newFont("assets/fonts/coolvetica.ttf", 24)
  self.santa_walk = Assets.sprites.santa_walk
  self.santa_walk_width = self.santa_walk:getWidth() / SANTA_WALK_FRAMES

  -- Calculate the bounding box for "Play" text
  self.play_width = self.menu_font:getWidth(self.play)
  self.play_height = self.menu_font:getHeight()
  self.play_x = (love.graphics.getWidth() - self.play_width) / 2.0
  self.play_y = (love.graphics.getHeight() - self.play_height) / 2.0 + 175
end

local function pointInRect(px, py, rx, ry, rw, rh)
  return px >= rx and px <= rx + rw and py >= ry and py <= ry + rh
end

function Instructions:quit()
  self.instructions_font:release()
end

function Instructions:update(dt) end

function Instructions:mousepressed(x, y, button)
  if button == 1 then
    if pointInRect(x, y, self.play_x, self.play_y, self.play_width, self.play_height) then
      local Game = require("states/game")
      GameStateManager:setState(Game:new())
    end
  end
end

function Instructions:draw()
  love.graphics.setColor(table.unpack(Assets.colors.white))
  love.graphics.draw(self.background, 0, 0)

  -- Draw the title
  love.graphics.setFont(self.font)
  love.graphics.print(self.title, (love.graphics.getWidth() - self.font:getWidth(self.title)) / 2.0, 10)

  -- Draw the instructions
  love.graphics.setColor(table.unpack(Assets.colors.background))
  love.graphics.rectangle("fill", 0, self.instructions_start_y - 25, love.graphics.getWidth(), 200)

  love.graphics.setColor(table.unpack(Assets.colors.white))
  love.graphics.setFont(self.instructions_font)
  self.instructions_font:setFilter("nearest", "nearest")
  love.graphics.print(
    "Hold down H or O or hold down on a touchscreen to make santa run. HO, HO, HO!",
    10,
    self.instructions_start_y
  )
  love.graphics.print(
    "Release the key or remove your finger from the touchscreen to stop him.",
    10,
    self.instructions_start_y + 40
  )
  love.graphics.print(
    "Stop as close as possible to the candy cane, but don't touch it or you know what happens!",
    10,
    self.instructions_start_y + 80
  )
  love.graphics.print(
    "The closer you get to the candy cane, the better, so try for your lowest score!",
    10,
    self.instructions_start_y + 120
  )

  -- Draw santa
  love.graphics.draw(
    self.santa_walk,
    love.graphics.newQuad(
      self.santa_walk_width * 4,
      0,
      self.santa_walk_width,
      self.santa_walk:getHeight(),
      self.santa_walk:getDimensions()
    ),
    0,
    love.graphics.getHeight() - self.santa_walk:getHeight() - 10,
    0,
    1
  )

  -- Draw play button
  love.graphics.setFont(self.menu_font)
  love.graphics.setColor(table.unpack(Assets.colors.white))
  love.graphics.print(self.play, self.play_x, self.play_y)
end

return Instructions
